### ISSUE TITLE
[css-text] letter-spacing should not apply to the end edge of a line/span?

### ISSUE THREAD

**Issue by kojiishi** [w3c/csswg-drafts#1518]
The spec defines so, but some tests http://output.jsbin.com/mobaxex/ indicates no impls do this today.

While what the spec says looks reasonable in many cases, this does not look web-compatible, and conformant behavior may look worse if authors have put workaround, or in some use cases where it's not desired. Should we make the new behavior opt-in?

Edge platform data usage indicates [44%](https://developer.microsoft.com/en-us/microsoft-edge/platform/usage/css/letter-spacing/) and Chrome usage indicates [30%](https://www.chromestatus.com/metrics/css/timeline/popularity/84).

**Comment by dbaron** [comment]
What leads to the conclusion that it doesn't look web-compatible?  I think in many cases the appearance that results is undesirable.

**Comment by kojiishi** [comment]
Can't agree more it's undesirable in most cases, I'm just saying this is not compatible with any existing browsers.

What if author has already put a workaround? Such as:
```css
span {
  letter-spacing: 1em;
  padding-left: 2em;
  padding-right: 1em;
  background-color: lime;
}
```
This renders good with all existing browsers but will be undesirable with conformant browser, correct?

**Comment by SebastianZ** [comment]
There is a long-standing related [Mozilla bug](https://bugzilla.mozilla.org/show_bug.cgi?id=125390).

For what it's worth, in order to different use cases I [suggested earlier](https://lists.w3.org/Archives/Public/www-style/2016Feb/0018.html) to extend the syntax of `letter-spacing` to allow to control whether the spacing should extend the text. The proposed syntax was:

    letter-spacing: [ normal | <length> ] [ inside || [ outside | inline-start | inline-end ] ]?

For web compatibility, the default value would then be `normal inside inline-end`. The value currently defined by the spec. would be `normal inside`.

Sebastian

**Comment by kojiishi** [comment]
Thank @SebastianZ for the info, this is a great help.

According to the [Mozilla bug 125390](https://bugzilla.mozilla.org/show_bug.cgi?id=125390), IE6 behaved like what the spec says, but Edge does the same as Gecko/Blink/WebKit. @atanassov do you happen to know the background?

I agree with you that if we want to change the behavior, I think we need to extend the syntax. I don't have clear idea which values we should define yet. I need better understanding on what the spec is requiring (#1509), and need to understand what is the good letter-spacing for bidi (#1517).

**Comment by kojiishi** [comment]
Blink bug is here FYI [crbug.com/267056](http://crbug.com/267056).

Its [comment 13](http://crbug.com/267056#c13) indicates that authors may put a workaround by applying negative right margin.

**Comment by kojiishi** [comment]
@SebastianZ, I support the idea to add a feature to switch the behavior. Two questions:
1. Would adding a keyword safe for serialization? If not, are you happy to have another property to control the behavior?
2. #1517 indicates a) existing behavior is not interoperable and b) no browsers do `inline-end`. Would you be ok with like `auto` for compat, or do we want to align the behavior first in #1517?

Not sure if @dbaron is happy to make this opt-in, or he still wants to change the default behavior.

@FremyCompany @litherum thoughts?

**Comment by fantasai** [comment]
I would prefer to not complicate letter-spacing, if that's at all possible, by adding unnecessary switches. Usage of letter-spacing is high, but that does not mean that a behavior change would regress websites: it's entirely possible that it would improve the situation. If we are concerned about web-compat, we should have some data indicating that the costs outweigh the benefits, and so far I don't see that here.

**Comment by kojiishi** [comment]
In [crbug.com/267056](http://crbug.com/267056), the commenter said he's using negative margins to solve the problem, and that this change will break his site. This is one proof, but from what you wrote, it looks to me that you're saying, you acknowledge the compat problem exists. Just saying it's small enough not to care, correct?

Can you show why you think this breaking change is ignorable?

**Comment by kojiishi** [comment]
and I'm also interested in hearing what other impls think. @litherum @FremyCompany ?

**Comment by FremyCompany** [comment]
I have mixed feelings about this. It is unfortunate that browser engines have settled on a sub-par design for the feature but I don't doubt that fixing letter-spacing (to leave the last letter of the line unaffected) wouldn't be trivial and would cause compat issues. 

That being said, I share @fantasai's point of view that adding a flag to letter-spacing is adding complexity to the platform that might not be worth the gain. The work-around (negative margin) for this particular issue doesn't seem very difficult to use when needed, so I would posit it is probably fine to keep things as-is in this case. 

Except if your language uses ideographic-like or monospace characters, letter-spacing is a not a functioning concept to start with:

![image](https://user-images.githubusercontent.com/364405/27492058-a263cfea-57f9-11e7-8d2b-0bf230001844.png)

**Comment by kojiishi** [comment]
Thank you @FremyCompany, that's a great idea I completely missed it exists. I'm good if we can agree with your proposal.

So, in high-level view, we have 3 options:
1. Define a new, better behavior without switches.
2. Add a new, better behavior as opt-in.
3. Go with existing behavior. Authors have workaround, not pretty but workable.

I'm good with 2 or 3.

If consensus is on 1 (I guess there is/was, correct?) I'd like to learn what we discussed from web compat perspective.

**Comment by aaaxx** [comment]
The problem with the negative-margin hack is that it doesn't help when the span also has a coloured background, so the third option is just not good enough.

By the way, the ligature handling is wrong in @FremyCompany's picture: they should get automatically disabled as soon as any amount of `letter-spacing` is applied.

**Comment by FremyCompany** [comment]
@aaaxx well, the handling of ligatures isn't wrong because font-feature-settings beats other properties <a href="https://drafts.csswg.org/css-fonts-4/#feature-variation-precedence">per the spec</a> but, you are right, if I didn't do anything, ligatures would have been disabled.

You are also right the workaround isn't perfect but the question is whether it is good enough for the desired use cases. Having a background on the inlines isn't working well in that scenario. Underlines is another example of something that doesn't work well. 

It would be interesting to study why people use letter-spacing on actual websites. I'm wondering, because even on Google.com, it is used as a hack for various things none of them striking me as the intended purpose of letter-spacing. Given how fragile hacks usually are and how widespread their use seems to be, changing the behavior of letter-spacing now seems unrealistic to me.

Another option to consider, if we decide the statu quo isn't good enough, would be to add a new property altogether. I would prefer to adding a flag for letter-spacing. We could repurpose letter-spacing to be exactly the behavior websites depend on today, then add glyph-spacing to behave as we wish letter-spacing should have worked.

**Comment by SebastianZ** [comment]
> 1. Would adding a keyword safe for serialization? If not, are you happy to have another property to control the behavior?

My earlier statement about `normal inside inline-end` being the default value was wrong. `normal` would stay the default, of course.

And serialization wouldn't be a problem, because both `normal` and `<length>` values (without the new keywords) would still be serialized the same way like now.

> 2. #1517 indicates a) existing behavior is not interoperable and b) no browsers do inline-end. Would you be ok with like `auto` for compat, or do we want to align the behavior first in #1517?

I just realized that my proposed syntax is web-incompatible. It should have been:

    letter-spacing: normal | <length> [ inside || [ outside | inline-start | inline-end ] ]?

Where `<length>` values without keywords are interpreted as `<length> inside inline-end`, for compatibility.

Regarding your question, I'd say the behavior should be aligned first, so it's easier to add the new values.

Also, I assume there may be no use case for having `inside` as separate value. I just added it for completeness. And `outside` could also be expressed by allowing both `inline-start` and `inline-end` to be specified at the same time. So, the syntax could be simplified to:

    letter-spacing: normal | <length> [ inline-start || inline-end ]?

Btw. my [use case](https://bugzilla.mozilla.org/show_bug.cgi?id=125390#c38) (having number input where each digit is displayed in its own "block" [copying the layout of a printed form]) would already be covered if browsers implemented the spec. as it's currently specified, i.e. if the spacing weren't applied at the end edge.
Though because that's not the case, I came up with this new syntax.

Sebastian

**Comment by kojiishi** [comment]
@SebastianZ:
> Regarding your question, I'd say the behavior should be aligned first, so it's easier to add the new values.

Great! Can we try this in #1517?

**Comment by kojiishi** [comment]
@aaaxx:
> The problem with the negative-margin hack is that it doesn't help when the span also has a coloured background, so the third option is just not good enough.

Agreed. We have two workarounds:
a. Use negative margin to eliminate the end-spacing.
b. Use positive padding to add the same spacing to the begin-side.

The workaround b works good if you need spacing on the begin side too, even when it has borders or background color.

Do you happen to know any real pages that has 1) letter-spacing, 2) has background-color or borders, and 3) don't want begin-side spacing? My imagination can't tell me how often this combination occurs, I hope you can help us on this.

**Comment by SebastianZ** [comment]
> @SebastianZ:
> 
>     Regarding your question, I'd say the behavior should be aligned first, so it's easier to add the new values.
> 
> Great! Can we try this in #1517?

Yes, that's what I wanted to say with that sentence. 😄 Sorry for not mentioning #1517 again!

> Do you happen to know any real pages that has 1) letter-spacing, 2) has background-color or borders, and 3) don't want start-side spacing?

My use case came up from a website, unfortunately not a public one. It is one allowing users to enter something restricted to a specific number of digits or characters. The form was basically copying a printed form used for tax declarations.

Sebastian

**Comment by tabatkins** [comment]
Post topic-closing resolution: No change to spec until we get data from Safari/Firefox trying it out.

**Comment by jonjohnjohnson** [comment]
After reading the current draft and going through this discussion and [[css-text] letter-spacing and word-spacing applied to which side? #1517](https://github.com/w3c/csswg-drafts/issues/1517) (as well as https://github.com/w3c/csswg-drafts/issues/1509 & https://github.com/w3c/csswg-drafts/issues/2165), I still can't decipher if browsers are suppose to implement A or B layout of this image. Combining a large `letter-spacing` value with different values of `text-align` regardless of `direction` or `unicode-bidi`.

![letter-spacing and text-align](https://user-images.githubusercontent.com/181364/37149366-4da8719e-2293-11e8-9fc3-00424a9b633f.png)
Notice the blue box outlining the extra space created by `letter-spacing`.

**Comment by kojiishi** [comment]
Thank you for asking. The spec says A, the 4 implementations do B.

**Comment by fantasai** [comment]
@jonjohnjohnson The spec very clearly states “Specifies additional spacing _between_ typographic character units” and “Letter-spacing must not be applied at the beginning or at the end of a line”. Could you explain where your confusion comes from so that we can fix that aspect of the spec?

**Comment by jonjohnjohnson** [comment]
@fantasai I guess, if I solely took the language in the current draft of the spec, I would have confidence in "A" being the recommendation, but was thrown off by the discussion here and elsewhere, primarily on the topic of web compatibility (i've used margin/padding to offset myself). So just hoping impls will be happy to give us this/a solution to well balanced typographic "tracking".

**Comment by fantasai** [comment]
Closing as no change per https://github.com/w3c/csswg-drafts/issues/1518#issuecomment-320197568

**Comment by kojiishi** [comment]
Re-opening, as no one has successfully shipped this yet after ~3 years, @fantasai and I are discussing how to move this forward offline.

**Comment by faceless2** [comment]
We've implemented letter spacing as specified, although are not yet shipping. Spacing between typographic character units seems to be the only sensible option when bidi is involved - with a more naive definition, simply adding letter-spacing after a letter, it all goes wrong very quickly.

Florian merged a [PR](https://github.com/web-platform-tests/wpt/pull/21602) from us a couple of months back for some tests demonstrating this:

http://wpt.live/css/css-text/letter-spacing/letter-spacing-bidi-003.xht
http://wpt.live/css/css-text/letter-spacing/letter-spacing-bidi-004.xht
http://wpt.live/css/css-text/letter-spacing/letter-spacing-bidi-005.xht

**Comment by kojiishi** [comment]
@faceless2 great you're going to ship it, look forward to the feedback. Each impl should have their own criteria how much they can accept the breakage.

Updates on impl statuses:

- For Blink, the same as before; the breakage is beyond what we can accept.
- @litherum wanted to give it a try in 2017, but told me that it won't happen in near future.
- @dbaron looks like supportive (from [Gecko 125390](https://bugzilla.mozilla.org/show_bug.cgi?id=125390))
- @fantasai thought the breakage is acceptable for Gecko in 2017, but when we went through a few real site examples, IIUC she seemed to understand the breakage is more than she thought. No conclusions yet.

A few URLs @fantasai and I read and/or examined together:
- http://artequalswork.com/posts/manual-kerning/
- https://designshack.net/articles/typography/kerning-live-web-type-with-jquery-and-lettering-js/
- http://kaoriha.org/dtp/tsume/ (this site sometimes uses margin but sometimes letter-spacing, see `tsume-a4` class)
- https://torounit.com/blog/2012/10/23/1372/ (in Japanese, please use your favorite translation service to read "it is widely known technique to use letter-spacing to remove spaces between inline-blocks)
- https://davidwalsh.name/remove-whitespace-inline-block#comment-61052 explains Yui3.Grid uses the technique mentioned above.

**Comment by kojiishi** [comment]
I think the miscommunication in the previous discussion was that, I think I somehow mislead people to believe that center/right-aligned case was the only case, when I actually wanted to say it was one of many cases the spec behavior will break.

**Comment by karip** [comment]
Note that changing this would affect [SVG2 letter-spacing and word-spacing](https://svgwg.org/svg2-draft/text.html#LetterSpacingProperty), because SVG2 refers to CSS Text Level 3 for letter-spacing and word-spacing definitions.

Currently, Chrome and Webkit render SVG text correctly without extra space at the end of the line. Firefox differs.

I haven't tested how browsers implement bidi text and other special cases for SVG.

Here's an [example](https://jsfiddle.net/0zd58omy/) of right aligned text in SVG and browser results:

![svg-letter-spacing](https://user-images.githubusercontent.com/6185066/81142301-fc009c00-8f77-11ea-9268-2a2716b051b7.png)

**Comment by kojiishi** [comment]
@karip Thank you for the info, this is very much helpful. If WG agrees to add a property or value to switch the behavior, we could say its initial value should do this for CSS and do that for SVG. That text can go to CSS Text or SVG. I look forward to hearing what SVG experts think we should do.

**Comment by fantasai** [comment]
It seems to me that the cases @kojiishi posted are largely requiring letter-spacing at the end of a span, in the middle of a line. They're not needing the letter-spacing at the end of the line. Letter-spacing at the end of the line is also the most problematic for the SVG case (and for CSS cases, too).

So my suggestion here is that:
    * We define for CSS Text Level 3 that letter-spacing is not applied at the end of the line. This solves the SVG case, improves rendering in many CSS cases, and will not break at least the letter-spacing-as-kerning examples that Koji posted.
    * We mark as undefined how letter-spacing applies at element/fragment boundaries in the middle of the line for Level 3, possibly describing the behaviors under consideration as options for Level 4.
    * We explore in level 4 the problem of letter-spacing applying at element at element/fragment boundaries in the middle of the line, possibly introducing a switch if necessary.

Does this seem like a workable plan?

**Comment by mruncreative** [comment]
Let me join the discussion saying how it should work:
*before the first character half of the additional space is inserted
*after every character the additional space is inserted
*after the last character half of the additional space is inserted instead of the entire additional space
How it currently works:
*after every character the additional space is inserted

This is important because letter-spacing has traditionally been used in Germany to highlight words in text like italic text in English speaking countries.
That's why the space to the sides of a word in a text should be **equal**. See the picture below for how not to do it:
![html problem](https://user-images.githubusercontent.com/67873653/86567124-1997b500-bf6b-11ea-8e87-0d02b55ba9c5.png)

**Comment by faceless2** [comment]
I disagree that what you're describing is optimal - a half-space surrounding the text isn't desired in many situations, and if you consider nested inlines with different letter-spacing values, it would massively complicate your proposal. The existing specification doesn't do this. But, if it were implemented the behaviour you describe could still be achieved (with `letter-spacing:2em; padding-inline: 1em`)

The problem is not specifying the correct behaviour, it's that the existing spec is not implemented because a large body of websites have added workarounds to the behaviour as it's currently implemented, and these sites would break if the specified behaviour were implemented.

If I can paraphrase what I believe is suggested above, the short term plan is to fix the most obviously broken case where letter spacing is messing with the spacing at the end of the line. The mid-term plan is to find a way to for authors to opt-in to the "new" letter-spacing behaviour, to enable it for those that want it, but not breaking existing websites that don't.

**Comment by mruncreative** [comment]
Well, then give me the chance to correct myself. It should work like this:
*after every character the additional space is inserted
*after the last character no additional space is inserted

Do you agree on this? Can we get this over with quickly?
There is one thing I'm sure everyone agrees: the spacing on both sides of a spaced word is equal!

**Comment by mruncreative** [comment]
>The mid-term plan is to find a way to for authors to opt-in to the "new" letter-spacing behaviour, to enable it for those that want it, but not breaking existing websites that don't.
No, that would be way more complicated than removing the workarounds from the few pages that used workarounds in my opinion.

**Comment by faceless2** [comment]
If you check the current spec you'll see it specifies that spacing is applied _between_ letters, not after. "spacing on both sides of a spaced word" is zero, and therefore equal. So I think we're agreeing on that point at least.

**Comment by astearns** [comment]
> No, that would be way more complicated than removing the workarounds from the few pages that used workarounds in my opinion.

I agree it’s more complicated, but it’s what we have to deal with in making improvements to CSS. Almost all the pages that use the workarounds will never be updated, so we cannot break things behind us.

**Comment by mruncreative** [comment]
>I agree it’s more complicated, but it’s what we have to deal with in making improvements to CSS. Almost all the pages that use the workarounds will never be updated, so we cannot break things behind us.
Okay, then do it. You guys are talking about it for **three years** already.
You all seem to know how it should work and already have a consensus about making a new CSS attribute in order to not break compatibility. You just have to do it.
What will you call it? "real-letter-spacing" instead of "letter-spacing"?

**Comment by faceless2** [comment]
Lots of things are impossible, and I have had plenty of dreams I don't want to come true under any circumstances.

Personally, I quite like something like `letter-spacing: 1em gap` for the "new version" and `letter-spacing: 1em` for the old version, but as far as I'm aware the discussion on details hasn't started yet. But I'm confident it will happen, so keep an eye on this issue as it will be cross referenced when it does.

**Comment by kojiishi** [comment]
@mruncreative:
> You all seem to know how it should work and already have a consensus about making a new CSS attribute in order to not break compatibility.

We don't have a consensus yet unfortunately. Chromium's original proposal is to add a CSS property to not to break existing pages. @fantasai and a few others are against adding a property, instead want to change existing behavior by break existing pages at the level not allowed for us to implement. We're stuck there for 3 years unfortunately. Your opinion is very much appreciated to make progress on this issue.

p.s. If you haven't read this already, can you please have a look at [Code of Conduct](https://github.com/w3c/csswg-drafts/blob/master/CODE_OF_CONDUCT.md)? Sorry if you have read this already.

**Comment by kojiishi** [comment]
@mruncreative
BTW, you might find elsewhere already but you can get what you want by:
```css
em {
  font-style: initial;
  letter-spacing: 2em;
  margin-left: 1em;
  margin-right: -1em;
}
```
https://jsbin.com/menabat/edit?html,output

**Comment by mruncreative** [comment]
>what you want by:

I don't want it anymore. I rethought what I want as soon as I was told about what the spec says.
>Well, then give me the chance to correct myself. It should work like this:

I want you people to find a solution to implement what the spec says. I don't care whether it is the existing letter-spacing, a new property or an extra value `letter-spacing: 1em gap`.
Just solve it. It's three years. The longer you take the more complicated the websites and ebooks get and if "@fantasai and a few others" get their way the impact of the change will be larger if you take another three years.

However I'm completely neutral towards the solution you guys choose. I just want you to choose on one ;)

**Comment by frivoal** [comment]
> I just want you to choose on one

A way forward has been proposed in https://github.com/w3c/csswg-drafts/issues/1518#issuecomment-652750185. I suggest we resume the discussion about that, as I think it might break the deadlock and could allow us to make progress.

Specifically, I'd be very interested in hearing what @kojiishi, as well as @litherum and @jfkthame, think about it.

**Comment by kojiishi** [comment]
I don't think the new proposal is shippable for us due to the number of pages it breaks. Talked about it to @fantasai, but since she believes the block usage is incorrect, she said she wants to break those pages and let authors fix them. Unfortunately we can't do that in Chromium.

IIUC, @dbaron thinks Mozilla can try shipping even the previous one, so my hope at this moment is Mozilla to ship and get reverted. From our experience on `break-word`, I hope that can convince WG?

**Comment by kojiishi** [comment]
I found @dbaron is happy to accept contributions if any but doesn't have plans to implement.

If that's still the case, how about:
* Make it undefined in level 3.
* Add a switch to level 4 (Did I understand correctly that @fantasai is ok to add a switch to level 4?)

Is this agreeable?

**Comment by kojiishi** [comment]
Started thinking about the "switch" in level 4, but then I found this is more complicated than I originally thought, because we don't know which character comes to the edge until we run the bidi algorithm, but that is after line break.

One possible idea is to compute line breaking by adding to the right of every character/atomic inline. Then after lines are broken and applied the bidi algorithm, but before computing alignment/justification, remove the spacing from the right end of each `letter-spacing` run. This can make the line overflow the available width if large negative `letter-spacing` is used, but probably acceptable because glyphs look overflowing anyway in such case.

Maybe there could be other ideas, I'll think about it a little more.

**Comment by kojiishi** [comment]
[More tests on SVG behavior](https://jsbin.com/cemagom/edit?html,output) with underline, `filter`, `getBoundingClientRect`, `textLength`, and bidi.
![letter-spacing-svg-mac-2020-07-20](https://user-images.githubusercontent.com/1274377/87882559-a0ed2a00-ca3b-11ea-8955-24a256ad1d40.png)

**Comment by kojiishi** [comment]
From the discussions above, here's the proposal for the level 4 "switch":
```css
letter-spacing-behavior: auto | between | right
```
<dl>
<dt>auto</dt>
<dd>

The behavior is determined by the following conditions:
* If the resolved direction of the character or atomic inline is `rtl`, the behavior is undefined.
* For CSS, if the _`letter-spacing` run_ (see `between` below) includes the rightmost in-flow fragment of a wrapped line after the bidi reorder, the behavior is undefined.
* For SVG, if the _`letter-spacing` run_ is at the right edge of a SVG text after the bidi reorder, the used value is `between`.
* Otherwise the used value is `right`.
</dd>
<dt>between</dt>
<dd>

When computing line wrapping, this value is the same as `right`.

Compute _`letter-spacing` runs_ after lines are wrapped and the bidi reorder was applied, but before computing alignment and justifications. A _`letter-spacing` run_ is a sequence of inline fragments that have the same used value of the `letter-spacing` property. Then for each _`letter-spacing` run_, remove the rightmost spacing.
</dd>
<dt>right</dt>
<dd>
Add the spacing to the right of every character or atomic inline.
</dd>
</dl>

Note, this `between` behavior may make lines to underflow if positive `letter-spacing`, or overflow if negative `letter-spacing`. I think this is acceptable because when `letter-spacing` is negative, glyphs may look overflowing anyway. The `between` behavior only changes how much they overflow.

**Comment by chrishtr** [comment]
I agree with @kojiishi that we need to have an opt-in for this feature due to risk of compatibility. I don't think the implementation complexity is all that high.

**Comment by fantasai** [comment]
> Did I understand correctly that @fantasai is ok to add a switch to level 4?

Only if we are sure it's absolutely necessary. I'm not yet so sure. :) But I think we can explore that option for L4, yes.

**Comment by chrishtr** [comment]
> Only if we are sure it's absolutely necessary.

Given the [very high usage](https://chromestatus.com/metrics/css/popularity#letter-spacing) and difficulty of estimating how many and how sites will break, we need this new switch to to opt into the desired behavior.

**Comment by jfkthame** [comment]
Personally, I've never been convinced that the current design/spec of letter-spacing as inserting space *between* typographic units is the right one, mostly because of the complications this introduces at boundaries. I think a much simpler model would be that letter-spacing increases (or decreases) the advance width of all typographic units equally on both sides, so that the glyph remains centered (typically) within its advance. Or in other words, half of the letter-spacing is applied to each of the left and right sidebearings of the glyph.

In particular, this works well for use-cases such as German-style  e m p h a s i s  where it's desirable for the letter-spacing to not only add some space *between* each letter of `<em>emphasis</em>` but also to adjust the inter-word spaces before and after it. With letter-space as currently spec'd, this could be done by adding margin-inline-start and -end (or with current implementations only adding it to the start), but with the alternative model we'd automatically get half the specified letter-spacing as an addition to the word spaces. This (IMO) would be much more convenient for authors.

For cases where (half) spacing at the inline-start and/or -end edge is *not* desired, it'd still be possible to suppress this by using a negative margin.

This also automatically works well at element boundaries within the line. And the symmetry means it's blind to bidi issues, unlike current behavior.

I think this would make for a simpler and more understandable model for both authors and implementers. But if -- as is quite likely -- changing the behavior of letter-spacing carries too much webcompat risk, perhaps we could introduce it as a separate property `tracking`, specified in terms of a simple, consistent effect on glyph sidebearings.

**Comment by kojiishi** [comment]
@jfkthame's idea works for me, it solves many issues including center-aligned case.

A few issues we'd like to keep in mind. It doesn't solve right-aligned case, but authors can still work around by using negative margins, and it is less common than center-aligned. It also doesn't solve excess underlines, which is [a little harder to work around](https://iamsteve.me/blog/entry/remove-letter-spacing-from-last-letter#removing-the-excess-underline-from-letter-spacing). Technically speaking, border/background too, but I think it is common to add spacing for these cases, so they might be ok.

I'm fine to have a switch or a new separate property, just need it to be opt-in.

**Comment by mruncreative** [comment]
@jfkthame
>I think a much simpler model would be that letter-spacing increases (or decreases) the advance width of all typographic units equally on both sides, so that the glyph remains centered (typically) within its advance. Or in other words, half of the letter-spacing is applied to each of the left and right sidebearings of the glyph.

This is exactly what I proposed further up in the discussion. However it is not necessary because you can simply include the spaces in the tag: `word<letter-spacing> word <letter-spacing>word`
We don't need special treatment in that regard.
As a German I care more about getting an extra spacing **html tag** with correct letter spacing see https://github.com/whatwg/html/issues/5709

**Comment by mruncreative** [comment]
Upon further research it also seems that also seems to be what typesetters did originally.

**Comment by mruncreative** [comment]
Letter spacing in German usage is basic formatting like underlining. So it's really important to be part of the content and transferable from one document to another and independent of CSS. As well as easily supportable by e-readers or the like.

**Comment by jfkthame** [comment]
I don't believe German usage requires a new tag; the appropriate HTML tag is `<em>`, which you can style with letter-spacing instead of italics when desired.

**Comment by mruncreative** [comment]
I can also style `<i>` or `<b>` but that's not the point. I want an exclusive tag for letter spacing.

**Comment by mruncreative** [comment]
Just like Americans have their exclusive tag for italic text no one here even used before the end of the war.

**Comment by mruncreative** [comment]
CSS styled tags are simply not a viable solution for archiving.

**Comment by mruncreative** [comment]
Or keeping the styling when the section is taken out of the original document.

**Comment by mruncreative** [comment]
Scrapers will probably also ignore it.

**Comment by jfkthame** [comment]
This is not the right place to argue about new HTML tags, I'm not going to engage further in that.

**Comment by jfkthame** [comment]
Regarding the actual issue here of how letter-spacing behaves: An alternative to a new `tracking` property would be to extend `letter-spacing` to take two values, which would be the extra space to be added on the left- and right-hand sides (or inline-start and inline-end sides; I'm not sure whether physical or logical is better here) of each typographic unit; so while current implementations of (single-valued) `letter-spacing` add all the space on one side, the two-valued syntax could be used to add half the desired space on each side, so that centering etc works nicely.

**Comment by mruncreative** [comment]
@jfkthame 
The first two were completely on topic. Can you please not just mark my words as "not to your liking" and thus hidden for everyone who does not have an account?
These are my first three comments for everyone to read:

mruncreative 1 hour ago
>I think a much simpler model would be that letter-spacing increases (or decreases) the advance width of all typographic units equally on both sides, so that the glyph remains centered (typically) within its advance. Or in other words, half of the letter-spacing is applied to each of the left and right sidebearings of the glyph.

>This is exactly what I proposed further up in the discussion. However it is not necessary because you can simply include the spaces in the tag: word<letter-spacing> word </letter-spacing>word
We don't need special treatment in that regard.
>As a German I care more about getting an extra spacing html tag with correct letter spacing see whatwg/html#5709

mruncreative 1 hour ago
>Upon further research that also seems to be what typesetters did originally.

mruncreative 1 hour ago
>Letter spacing in German usage is basic formatting like underlining. So it's really important to be part of the content and transferable from one document to another and independent of CSS. As well as easily supportable by e-readers or the like.

**Comment by fantasai** [comment]
@mruncreative Your comments are off-topic because we are not discussing HTML tags here. I have hidden them, so that they do not distract from the on-topic discussion here. If you continue with your off-topic comments I will delete them.

**Comment by psnet** [comment]
Want to note this "feature" with buggy nature is still actual, gap in details for implementation makes situation when devs will fix this by negative margin and later this will lead to broken design after standard will change to space **around** letters
